pkg/helm/client.go

```diff
@@ -0,0 +1,247 @@
+	v1, err := semver.Make(rc[i].Version)
+	if err != nil {
+		return false
+	}
+	v2, err := semver.Make(rc[j].Version)
+	if err != nil {
+		return false
+	}
+
+	return v1.Compare(v2) == 1
+}
+
+type Getter interface {
+	Get(string) (*http.Response, error)
+}
+
+type getter struct {
+	client *http.Client
+}
+
+func newGetter() *getter {
+	c := &http.Client{
+		Timeout: 30 * time.Second,
+	}
+
+	return &getter{
+		client: c,
+	}
+}
+
+func (g *getter) Get(s string) (*http.Response, error) {
+	req, err := http.NewRequest(http.MethodGet, s, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return g.client.Do(req)
+}
+
+// HTTPClient is a HTTP Helm repository client.
+type HTTPClient struct {
+	url    *url.URL
+	getter Getter
+}
+
+var _ RepositoryClient = (*HTTPClient)(nil)
+
+// NewHTTPClient creates an instance of HTTPClient
+func NewHTTPClient(urlStr string, hg Getter) (*HTTPClient, error) {
+	normalized, err := normalizeHelmURI(urlStr)
+	if err != nil {
+		return nil, errors.Wrap(err, "normalizing Helm repository URL")
+	}
+
+	u, err := url.Parse(normalized)
+	if err != nil {
+		return nil, err
+	}
+
+	if hg == nil {
+		hg = newGetter()
+	}
+
+	return &HTTPClient{
+		url:    u,
+		getter: hg,
+	}, nil
+}
+
+// Repository returns the Helm repository's content.
+func (hrc *HTTPClient) Repository() (*Repository, error) {
+	resp, err := hrc.getter.Get(hrc.url.String())
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving repository index.yaml response")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("repository returned unexpected response %s when retrieving %q",
+			resp.Status, hrc.url.String())
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading repository index.yaml body")
+	}
+
+	var hre Repository
+
+	if err := yaml.Unmarshal(b, &hre); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling repository index.yaml")
+	}
+
+	return &hre, nil
+}
+
+// Chart returns a chart from the repository. If version is blank, it returns the latest.
+func (hrc *HTTPClient) Chart(name, version string) (*RepositoryChart, error) {
+	repo, err := hrc.Repository()
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving repository")
+	}
+
+	if version == "" {
+		for _, chart := range repo.Latest() {
+			if name == chart.Name {
+				return &chart, nil
+			}
+		}
+
+		return nil, errors.Errorf("chart %q was not found", name)
+	}
+
+	charts, ok := repo.Charts[name]
+	if !ok {
+		return nil, errors.Errorf("chart %q was not found", name)
+	}
+
+	for _, chart := range charts {
+		if version == chart.Version {
+			return &chart, nil
+		}
+	}
+
+	return nil, errors.Errorf("chart %q with version %q was not found", name, version)
+}
+
+// Fetch fetches URLs from a repository. If uri is a path, it will use the client URL as the base.
+func (hrc *HTTPClient) Fetch(uri string) (io.ReadCloser, error) {
```

Comment:
Should this be exported?

Reply:
Yes, because the Helm registry uses it to resolve files and it is in another package.